nupic/research/frameworks/greedy_infomax/models/ResNetEncoder.py

```diff
@@ -0,0 +1,767 @@
+            )
+
+    def forward(self, x):
+        out = self.nonlinearity1(x)
+        shortcut = self.shortcut(out) if hasattr(self, "shortcut") else x
+        out = self.conv1(out)
+        out = self.nonlinearity1(out)
+        out = self.conv2(out)
+        out = self.nonlinearity3(out)
+        out = self.conv3(out)
+        out += shortcut
+        return out
+
+
+class VDropSparsePreActBottleneckNoBN(nn.Module):
+    """Pre-activation version of the original Bottleneck module."""
+
+    expansion = 4
+
+    def __init__(self, in_planes, planes, stride=1, percent_on=None, central_data=None):
+        super(VDropSparsePreActBottleneckNoBN, self).__init__()
+        if percent_on is None:
+            percent_on = {
+                "nonlinearity1": 0.9,
+                "nonlinearity2": 0.9,
+                "nonlinearity3": 0.9,
+            }
+
+        self.conv1 = VDropConv2d(
+            in_planes, planes, kernel_size=1, central_data=central_data
+        )
+        self.conv2 = VDropConv2d(
+            planes,
+            planes,
+            kernel_size=3,
+            central_data=central_data,
+            stride=stride,
+            padding=1,
+        )
+        self.conv3 = VDropConv2d(
+            planes, self.expansion * planes, kernel_size=1, central_data=central_data
+        )
+        if stride != 1 or in_planes != self.expansion * planes:
+            self.shortcut = nn.Sequential(
+                VDropConv2d(
+                    in_planes,
+                    self.expansion * planes,
+                    kernel_size=1,
+                    central_data=central_data,
+                    stride=stride,
+                )
+            )
+        # activation sparsity
+        if percent_on["nonlinearity1"] >= 0.5:
+            self.nonlinearity1 = F.relu
+        else:
+            self.nonlinearity1 = KWinners2d(
+                in_planes, percent_on=percent_on["nonlinearity1"]
+            )
+        if percent_on["nonlinearity2"] >= 0.5:
+            self.nonlinearity1 = F.relu
+        else:
+            self.nonlinearity1 = KWinners2d(
+                in_planes, percent_on=percent_on["nonlinearity2"]
+            )
+        if percent_on["nonlinearity3"] >= 0.5:
+            self.nonlinearity1 = F.relu
+        else:
+            self.nonlinearity1 = KWinners2d(
+                in_planes, percent_on=percent_on["nonlinearity3"]
+            )
+
+    def forward(self, x):
+        out = self.nonlinearity1(x)
+        shortcut = self.shortcut(out) if hasattr(self, "shortcut") else x
+        out = self.conv1(out)
+        out = self.nonlinearity1(out)
+        out = self.conv2(out)
+        out = self.nonlinearity3(out)
+        out = self.conv3(out)
+        out += shortcut
+        return out
+
+
+class ResNetEncoder(nn.Module):
+    """
+    The main subcomponent of FullVisionModel. This encoder also implements both
+    .forward() and .encode() to support different outputs for unsupervised and
```

Comment:
It turns out there's already some implementation of ResNets (or maybe just parts of them) in nupic.research. Check out the [pytorch folder](https://github.com/numenta/nupic.research/tree/master/nupic/research/frameworks/pytorch/models). Is there anything there you can re-use for any of your model implementation? We should avoid duplicate code wherever possible

Reply:
I thought about this when I was implementing the codebase. Yes, it is possible to go back and use parts of the current codebase but I decided against it because:
1) Code reproducibility. Maybe we implemented our ResNet in a slightly different way than the GIM paper, which would not lead to comparable results
2) Instead of using a BasicBlock, the paper uses a modified version of the block with no batch norm, which we do not have in resnets.py
3) The ResNet class includes both the NoBatchnormBlock and also the BilinearInfo block. In order to use the BilinearInfo like the paper, I would have to create a new class anyway

It's still possible to reuse a bit of the code, but it will take a bit of time to make sure that the implementations are comparable again. Plus, we have now moved over to the BlockModel. I'm trying to walk the line between codebase maintenance and moving forward with the goals of the project. Open to suggestions.